Marlin/src/HAL/HAL_STM32/timers.cpp

```diff
@@ -32,62 +32,89 @@
 
 #define NUM_HARDWARE_TIMERS 2
 
+#define __TIMER_DEV(X) TIM##X
+#define _TIMER_DEV(X) __TIMER_DEV(X)
+#define STEP_TIMER_DEV _TIMER_DEV(STEP_TIMER)
+#define TEMP_TIMER_DEV _TIMER_DEV(TEMP_TIMER)
+
 // ------------------------
 // Private Variables
 // ------------------------
 
-stm32_timer_t TimerHandle[NUM_HARDWARE_TIMERS];
+HardwareTimer *timer_instance[NUM_HARDWARE_TIMERS] = { NULL };
+bool timer_enabled[NUM_HARDWARE_TIMERS] = { false };
 
 // ------------------------
 // Public functions
 // ------------------------
 
-bool timers_initialized[NUM_HARDWARE_TIMERS] = { false };
-
+// frequency is in Hertz
 void HAL_timer_start(const uint8_t timer_num, const uint32_t frequency) {
+  static_assert(TIM_IRQ_PRIO > 0 && TIM_IRQ_PRIO < 15, "Default timer interrupt priority is too low to allow proper prioritization of interrupts.");
 
-  if (!timers_initialized[timer_num]) {
-    uint32_t step_prescaler = STEPPER_TIMER_PRESCALE - 1,
-                       temp_prescaler = TEMP_TIMER_PRESCALE - 1;
+  if (!HAL_timer_initialized(timer_num)) {
     switch (timer_num) {
-      case STEP_TIMER_NUM:
-        // STEPPER TIMER - use a 32bit timer if possible
-        TimerHandle[timer_num].timer = STEP_TIMER_DEV;
-        TimerHandle[timer_num].irqHandle = Step_Handler;
-        TimerHandleInit(&TimerHandle[timer_num], (((HAL_TIMER_RATE) / step_prescaler) / frequency) - 1, step_prescaler);
-        HAL_NVIC_SetPriority(STEP_TIMER_IRQ_NAME, STEP_TIMER_IRQ_PRIO, 0);
+      case STEP_TIMER_NUM: // STEPPER TIMER - use a 32bit timer if possible
+        timer_instance[timer_num] = new HardwareTimer(STEP_TIMER_DEV);
+        /* Set the prescaler to the final desired value.
+         * This will change the effective ISR callback frequency but when
+         * HAL_timer_start(timer_num=0) is called in the core for the first time
+         * the real frequency isn't important as long as, after boot, the ISR
+         * gets called with the correct prescaler and count register. So here
+         * we set the prescaler to the correct, final value and ignore the frequency
+         * asked. We will call back the ISR in 1 second to start at full speed.
+         *
+         * The proper fix, however, would be a correct initialization OR a
+         * HAL_timer_change(const uint8_t timer_num, const uint32_t frequency)
+         * which changes the prescaler when an IRQ frequency change is needed
+         * (for example when steppers are turned on)
+         */
+        timer_instance[timer_num]->setPrescaleFactor(STEPPER_TIMER_PRESCALE); //the -1 is done internally
+        timer_instance[timer_num]->setOverflow(_MIN(hal_timer_t(HAL_TIMER_TYPE_MAX), (HAL_TIMER_RATE) / (STEPPER_TIMER_PRESCALE) /* /frequency */), TICK_FORMAT);
```

Comment:
Note this change from a recent PR. The set value is no longer divided by `frequency`.